### ISSUE TITLE
Settings panel should allow opening to a subcategory

### ISSUE THREAD

**Issue by Meorawr** [Stanzilla/WoWUIBugs#285]
In Dragonflight the current incarnation of the settings APIs doesn't support directly opening the settings panel to any category that isn't a top-level one.

A test addon is provided that will register a tree of addon categories up to a depth of 3. Once the test addon is loaded, the command listed in the table below can be used to test it.

[Meorawr_SettingsOpenSubcategoryTest.zip](https://github.com/Stanzilla/WoWUIBugs/files/9638662/Meorawr_SettingsOpenSubcategoryTest.zip)

<details>
<summary>Meorawr_SettingsOpenSubcategoryTest.lua</summary>

```lua
local g_counter = CreateCounter();

local function CreateTestCategory(color, parent)
    local frame = CreateFrame("Frame");
    local background = frame:CreateTexture();
    background:SetAllPoints(frame);
    background:SetColorTexture(color:GetRGBA());

    local category;
    if parent then
        local name = string.format("%s.%d", parent:GetName(), parent.counter());
        category = Settings.RegisterCanvasLayoutSubcategory(parent, frame, name);
    else
        local name = string.format("Test Category %d", g_counter());
        category = Settings.RegisterCanvasLayoutCategory(frame, name);
    end

    category.counter = CreateCounter();

    Settings.RegisterAddOnCategory(category);
    return category;
end

TestCategory1 = CreateTestCategory(RED_FONT_COLOR);
TestCategory1_1 = CreateTestCategory(BLUE_FONT_COLOR, TestCategory1);
TestCategory1_1_1 = CreateTestCategory(GREEN_FONT_COLOR, TestCategory1_1);
TestCategory1_1_2 = CreateTestCategory(GRAY_FONT_COLOR, TestCategory1_1);
TestCategory1_2 = CreateTestCategory(YELLOW_FONT_COLOR, TestCategory1);
TestCategory1_3 = CreateTestCategory(ORANGE_FONT_COLOR, TestCategory1);
TestCategory1_3_1 = CreateTestCategory(BATTLENET_FONT_COLOR, TestCategory1_3);
TestCategory1_3_2 = CreateTestCategory(TRANSMOGRIFY_FONT_COLOR, TestCategory1_3);
TestCategory2 = CreateTestCategory(GOLD_FONT_COLOR);
TestCategory2_1 = CreateTestCategory(DIM_GREEN_FONT_COLOR, TestCategory2);
TestCategory2_1_1 = CreateTestCategory(AREA_NAME_FONT_COLOR, TestCategory2_1);
```

</details>

| Command | Outcome | Desired Result |
| -------------- | :--------- | :--|
| `/run Settings.OpenToCategory(TestCategory1:GetID())` | The settings panel successfully opens to the first custom category. | N/A |
| `/run Settings.OpenToCategory(TestCategory2:GetID())` | The settings panel successfully opens to the second custom category. | N/A |
| `/run Settings.OpenToCategory(TestCategory2_1:GetID())` | The settings panel fails to open to the requested top-level category. | The settings panel should open to the requested category by an ID. | 
| `/run Settings.OpenToCategory(TestCategory1_3_2:GetID())` | The settings panel fails to open to the requested nested subcategory. | The settings panel should open to the requested nested subcategory by an ID, and should recursively expand its parent hierarchy. |

[Settings.OpenToCategory]: https://github.com/Gethe/wow-ui-source/blob/c851b385acc54a08438ecbc639a9ae61be353c51/Interface/SharedXML/Settings/Blizzard_Settings.lua#L111-L113
[SettingsPanelMixin:OpenToCategory]: https://github.com/Gethe/wow-ui-source/blob/c851b385acc54a08438ecbc639a9ae61be353c51/Interface/SharedXML/Settings/Blizzard_SettingsPanel.lua#L213-L218
[SettingsCategoryListMixin:GetCategory]: https://github.com/Gethe/wow-ui-source/blob/c851b385acc54a08438ecbc639a9ae61be353c51/Interface/SharedXML/Settings/Blizzard_CategoryList.lua#L207-L211
[SettingsCategoryListMixin:SetCurrentCategory]: https://github.com/Gethe/wow-ui-source/blob/c851b385acc54a08438ecbc639a9ae61be353c51/Interface/SharedXML/Settings/Blizzard_CategoryList.lua#L260-L272

**Comment by DFortun81** [comment]
Bumping this. It would be really nice to have this.

**Comment by DFortun81** [comment]
A modification to the function like this would possibly do as we need it to for this to function.

```
local function FindCategoryByCategoryID(categories, categoryID)
	if categories then
		for index, category in ipairs(categories) do
			if category:GetID() == categoryID then
				return category;
			else
				local subcategory = FindCategoryByCategoryID(category:GetSubcategories(), categoryID);
				if subcategory then return subcategory; end
			end
		end
	end
end

function SettingsPanelMixin:OpenToCategory(categoryID, scrollToElementName)
	self:Open();

	local categoryTbl = FindCategoryByCategoryID(self:GetAllCategories(), categoryID);
	if not categoryTbl then return false; end
	
	self:SelectCategory(categoryTbl);
	if scrollToElementName then
		self:GetSettingsList():ScrollToElementByName(scrollToElementName);
	end
	return true;
end
```

**Comment by thewowcollector** [comment]
Blizzard should be willing to do anything they can to support Addon developers who allow players to enjoy the game more. They support the game for free, please help them out!

**Comment by KTom71** [comment]
Einfach einmal testen.
`/run Settings.OpenToCategory(); SettingsPanel:SelectCategory(TestCategory2_1, true)`

**Comment by KTom71** [comment]
Try it once
`/run SettingsPanel:Open(); SettingsPanel:SelectCategory(TestCategory2_1, true)`

**Comment by Arxareon** [comment]
> Try it once `/run SettingsPanel:Open(); SettingsPanel:SelectCategory(TestCategory2_1, true)`

It actually works wonderfully*! Is this new?

Thanks! 🍻

*_But taints the UI, see below._

**Comment by Meorawr** [comment]
> Try it once `/run SettingsPanel:Open(); SettingsPanel:SelectCategory(TestCategory2_1, true)`

The problem with this approach is that you taint the settings panel. Run the following, click the Keybindings category, then close the settings panel and you'll see the issue.

```
/run CAT = Settings.RegisterCanvasLayoutCategory(CreateFrame("Frame"), "My AddOn"); Settings.RegisterAddOnCategory(CAT); SettingsPanel:Open(); SettingsPanel:SelectCategory(CAT, true);
```

**Comment by Arxareon** [comment]
> > > Try it once `/run SettingsPanel:Open(); SettingsPanel:SelectCategory(TestCategory2_1, true)`
> 
> The problem with this approach is that you taint the settings panel. Run the following, click the Keybindings category, then close the settings panel and you'll see the issue.
> 
> ```
> /run CAT = Settings.RegisterCanvasLayoutCategory(CreateFrame("Frame"), "My AddOn"); Settings.RegisterAddOnCategory(CAT); SettingsPanel:Open(); SettingsPanel:SelectCategory(CAT, true);
> ```

~~I do not get taint popups when testing an implementation of the suggested approach. I will have to investigate deeper tomorrow.~~
Yes, yes it does taint, unfortunately.

> 6/9 17:30:27.613      Interface/AddOns/Blizzard_UIParent/Mainline/UIParent.lua:2807 ToggleGameMenu()
6/9 17:30:27.613      Interface/AddOns/Blizzard_Settings_Shared/Blizzard_SettingsPanel.lua:360 SettingsPanel:TransitionBackOpeningPanel()
6/9 17:30:27.613      Interface/AddOns/Blizzard_Settings_Shared/Blizzard_SettingsPanel.lua:349 SettingsPanel:ExitWithCommit()
6/9 17:30:27.613      Interface/AddOns/Blizzard_Settings_Shared/Blizzard_SettingsPanel.lua:318 SettingsPanel:Close()
6/9 17:30:27.613      Interface/AddOns/Blizzard_UIParent/Mainline/UIParent.lua:2785 ToggleGameMenu()
6/9 17:30:27.613      TOGGLEGAMEMENU:1

I tried fiddling around with `securecallfunction` but no luck.

**Comment by KTom71** [comment]
Why does Blizzard put something like this into the ToggleGameMenu() function?

`	elseif ( securecall("OrderHallMissionFrame_EscapePressed") ) then`
`	elseif ( securecall("OrderHallTalentFrame_EscapePressed") ) then`
`	elseif ( securecall("BFAMissionFrame_EscapePressed") ) then`
_**`	elseif ( SpellStopCasting() ) then`**_
_**`	elseif ( SpellStopTargeting() ) then`**_
`	elseif(MatchCelebrationPartyPoseFrame and MatchCelebrationPartyPoseFrame:IsShown()) then`
`	elseif ( SoulbindViewer and SoulbindViewer:HandleEscape()) then`
`	elseif ( ClassTalentFrame and ClassTalentFrame:IsShown() ) then`
`		ClassTalentFrame:CheckConfirmClose();`
`	elseif ( ProfessionsFrame and ProfessionsFrame:IsShown() ) then`
`		ProfessionsFrame:CheckConfirmClose();`

**Comment by Meorawr** [comment]
Partially resolved in 11.0.2.55789; last test case doesn't seem to expand the parent hierarchy - but it does fundamentally at least navigate to the category.
